CLOUDSTACK-8678: Reserve RAM for KVM host OS

kmccormick opened the pull request:
Use host.reserved.ram.mb agent property to modify total system RAM
before reporting to management server.

DaanHoogland commented on plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/resource/LibvirtComputingResource.java:
```diff
@@ -258,7 +258,7 @@
     private String _updateHostPasswdPath;
 
-    private int _dom0MinMem;
+    private long _dom0MinMem;
```
I think we need to initialize to a reasonable default or 0L. Not sure if it is possible by installation but otherwise the properties file might be corrupted or empty.

wido replied:
Agree, I think we should always reserve at least 1GB to keep a system alive. Admins can always set it so something lower.

securityvoodoo replied:
I see no reason not to default it to something sane.
Wido - 
The way it is setup right now, admins would not be able to change it to something lower than whatever we default to unless still greater than 768mb or 10% of system ram.  At least not by specifying it via the host.reserved.mem.mb property.  That being said, the default should be greater than 768mb.  A 1G default sounds sane to me.

DaanHoogland commented:
except for the initialisation issue, LGTM

kmccormick commented:
OK, reworked this a bit. Removed dom0ram and using only _dom0MinMem now.

This line should cause a 1GB default if the value isn't specified in the file:
_dom0MinMem = NumbersUtil.parseInt(value, 1024) * 1024 * 1024L;

No longer doing the odd min/max calculation, just using either the fixed default or the admin specified value.

DaanHoogland commented:
LGTM, in case my previous comment is not already interpreted as such.

koushik-das commented:
What happens in case of an upgrade? Looks like there is a possibility that the host may get over-provisioned. Should there be some scripts to let the user know if some VMs needs to be destroyed before upgrade?

Also do we need a similar behaviour for cpu?

kmccormick commented:
Regarding the upgrade scenario, I believe that this would not affect any running VMs. Can anyone confirm? Besides just running VMs, it is still a potential problem for an installation that is already near capacity, as the upgrade would remove a little bit of capacity. For both issues, I think a mention in the release notes would be sufficient. To keep the old behavior, admins just need to add host.reserved.ram.mb=0 to their agent.properties file.

CPU I don't think is an issue, as memory is typically the limiting factor in virtualization. If the CPU gets overly taxed, things slow down, where if all the memory gets used up, VMs or other processes get killed.

wido commented:
@kmccormick Agreed. A release note would be enough since running VMs will not be affected.

CPU is a different problem indeed. Stuff becomes slow, but going Out of Memory is a bigger problem.

remibergsma commented:
@kmccormick thanks for the PR, will test it soon! Had a bit of a rough week at the office with some emergencies going on. 
Please rebase with current master, as we cannot merge it with conflicts.

remibergsma commented:
LGTM Tested it, works great (even without single change to agent.properties). It will reserve 1GB by default.

![before]
(http://s29.postimg.org/uoxtrihhz/Screen_Shot_2015_09_04_at_22_51_02.png)

![with new agent from this PR]
(http://s15.postimg.org/rbin8ipff/Screen_Shot_2015_09_04_at_23_04_01.png)

Thanks for fixing this @kmccormick !

remibergsma commented:
No Travis test results? Both Apache builds are green and I verified the function to work properly so merging it.
